File: scripts/pack-fabric/routing/lib/region-polygons.js
"use strict";

const path = require("path");

const RECORD_PATH = path.join(__dirname, "..", "schema", "region-polygons.v1.json");
const RECORD = require("../schema/region-polygons.v1.json");

let cached = null;

function loadRegionPolygons() {
  if (cached) return cached;
  cached = RECORD;
  return cached;
}

function pointInRing(lon, lat, ring) {
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const xi = Number(ring[i][0]);
    const yi = Number(ring[i][1]);
    const xj = Number(ring[j][0]);
    const yj = Number(ring[j][1]);
    const intersect = yi > lat !== yj > lat && lon < ((xj - xi) * (lat - yi)) / (yj - yi + 0.0) + xi;
    if (intersect) inside = !inside;
  }
  return inside;
}

function pointInPolygonCoords(lon, lat, rings) {
  if (!Array.isArray(rings) || !rings[0] || !pointInRing(lon, lat, rings[0])) return false;
  for (let h = 1; h < rings.length; h += 1) {
    if (pointInRing(lon, lat, rings[h])) return false;
  }
  return true;
}

function pointInGeometry(lon, lat, geom) {
  if (!geom || !geom.type) return false;
  if (geom.type === "Polygon") return pointInPolygonCoords(lon, lat, geom.coordinates);
  if (geom.type === "MultiPolygon") {
    return (geom.coordinates || []).some((polygon) => pointInPolygonCoords(lon, lat, polygon));
  }
  return false;
}

function pointInRegionPolygon(regionId, lon, lat) {
  const id = String(regionId || "").toLowerCase();
  const geom = loadRegionPolygons().regions && loadRegionPolygons().regions[id];
  if (!geom) return false;
  return pointInGeometry(lon, lat, geom);
}

function geometryBbox(geom) {
  let minLon = Infinity;
  let minLat = Infinity;
  let maxLon = -Infinity;
  let maxLat = -Infinity;
  function walk(coords) {
    if (!Array.isArray(coords) || !coords.length) return;
    if (typeof coords[0] === "number") {
      const lon = Number(coords[0]);
      const lat = Number(coords[1]);
      if (lon < minLon) minLon = lon;
      if (lat < minLat) minLat = lat;
      if (lon > maxLon) maxLon = lon;
      if (lat > maxLat) maxLat = lat;
      return;
    }
    for (const child of coords) walk(child);
  }
  walk(geom && geom.coordinates);
  if (!Number.isFinite(minLon)) return null;
  return { minLon, minLat, maxLon, maxLat };
}

function bboxArea(box) {
  if (!box) return Infinity;
  return Math.max(0, box.maxLon - box.minLon) * Math.max(0, box.maxLat - box.minLat);
}

function expandBbox(box, padDeg) {
  return {
    minLon: box.minLon - padDeg,
    minLat: box.minLat - padDeg,
    maxLon: box.maxLon + padDeg,
    maxLat: box.maxLat + padDeg
  };
}

function intersectBboxes(a, b) {
  const box = {
    minLon: Math.max(a.minLon, b.minLon),
    minLat: Math.max(a.minLat, b.minLat),
    maxLon: Math.min(a.maxLon, b.maxLon),
    maxLat: Math.min(a.maxLat, b.maxLat)
  };
  if (box.minLon >= box.maxLon || box.minLat >= box.maxLat) return null;
  return box;
}

function seamCorridor(leftId, rightId, padDeg = 0.15) {
  const regions = loadRegionPolygons().regions || {};
  const left = geometryBbox(regions[String(leftId || "").toLowerCase()]);
  const right = geometryBbox(regions[String(rightId || "").toLowerCase()]);
  if (!left || !right) return null;
  const overlap = intersectBboxes(expandBbox(left, padDeg), expandBbox(right, padDeg));
  return overlap;
}

function pointInBbox(lon, lat, box) {
  return lon >= box.minLon && lon <= box.maxLon && lat >= box.minLat && lat <= box.maxLat;
}

/**
 * Admin-polygon owner. Null when the point is in none of the loaded polygons.
 * NS/NB overlap on the isthmus uses the meridian already proven in the field.
 */
function polygonOwner(lon, lat) {
  const regions = loadRegionPolygons().regions || {};
  const hits = [];
  for (const id of Object.keys(regions)) {
    const box = geometryBbox(regions[id]);
    if (box && !pointInBbox(lon, lat, box)) continue;
    if (pointInRegionPolygon(id, lon, lat)) hits.push(id);
  }
  if (!hits.length) return null;
  if (hits.length === 1) return hits[0];
  if (hits.includes("ns") && hits.includes("nb")) {
    return lon >= -64.27 ? "ns" : "nb";
  }
  const txPieces = hits.filter(id => id.startsWith("tx-")).sort();
  if (txPieces.length) {
    const owner = `tx-${lat >= 31 ? "n" : "s"}${lon >= -97.25 ? "e" : "w"}`;
    return txPieces.includes(owner) ? owner : txPieces[0];
  }
  // Ontario South/North share the cut edge; prefer the published half by latitude.
  const onHalves = hits.filter((id) => id === "on-s" || id === "on-n");
  if (onHalves.length) {
    return lat >= 46.0 ? (onHalves.includes("on-n") ? "on-n" : onHalves[0]) : (onHalves.includes("on-s") ? "on-s" : onHalves[0]);
  }
  hits.sort((a, b) => bboxArea(geometryBbox(regions[a])) - bboxArea(geometryBbox(regions[b])));
  return hits[0];
}

function maritimesOwner(lon, lat) {
  return polygonOwner(lon, lat);
}

function resetPolygonCache() {
  cached = null;
}

module.exports = {
  RECORD_PATH,
  geometryBbox,
  loadRegionPolygons,
  maritimesOwner,
  pointInBbox,
  pointInGeometry,
  pointInRegionPolygon,
  polygonOwner,
  resetPolygonCache,
  seamCorridor
};

File: Dirt/Routing/OnDevice/RegionPolygons.swift
import Foundation

/// OSM administrative polygons for stamped v3 regions.
/// Lockstep with `scripts/pack-fabric/routing/lib/region-polygons.js`.
enum RegionPolygons {
    private struct BBox {
        var minLon: Double
        var minLat: Double
        var maxLon: Double
        var maxLat: Double
        var area: Double { max(0, maxLon - minLon) * max(0, maxLat - minLat) }
        func contains(lon: Double, lat: Double) -> Bool {
            lon >= minLon && lon <= maxLon && lat >= minLat && lat <= maxLat
        }
    }

    private static let geometries: [String: [String: Any]] = {
        let url = Bundle.main.url(forResource: "RegionPolygons", withExtension: "json")
            ?? Bundle(for: BundleToken.self).url(forResource: "RegionPolygons", withExtension: "json")
        guard let url, let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let regions = root["regions"] as? [String: [String: Any]]
        else { return [:] }
        return regions
    }()

    private static let bboxes: [String: BBox] = {
        var out: [String: BBox] = [:]
        for (id, geom) in geometries {
            if let box = geometryBbox(geom) { out[id] = box }
        }
        return out
    }()

    static func contains(_ regionId: String, longitude lon: Double, latitude lat: Double) -> Bool {
        guard let geom = geometries[regionId.lowercased()],
              let type = geom["type"] as? String
        else { return false }
        if let box = bboxes[regionId.lowercased()], !box.contains(lon: lon, lat: lat) {
            return false
        }
        return pointInGeometry(lon: lon, lat: lat, type: type, coordinates: geom["coordinates"])
    }

    /// Admin-polygon owner. Nil when the point is in none of the loaded polygons.
    static func polygonOwner(longitude lon: Double, latitude lat: Double) -> String? {
        var hits: [String] = []
        for id in geometries.keys where contains(id, longitude: lon, latitude: lat) {
            hits.append(id)
        }
        if hits.isEmpty { return nil }
        if hits.count == 1 { return hits[0] }
        if hits.contains("ns"), hits.contains("nb") {
            return lon >= -64.27 ? "ns" : "nb"
        }
        let onHalves = hits.filter { $0 == "on-s" || $0 == "on-n" }
        if !onHalves.isEmpty {
            if lat >= 46.0, onHalves.contains("on-n") { return "on-n" }
            if lat < 46.0, onHalves.contains("on-s") { return "on-s" }
            return onHalves[0]
        }
        let qcHalves = hits.filter { $0 == "qc-s" || $0 == "qc-n" }
        if !qcHalves.isEmpty {
            if lat >= 49.0, qcHalves.contains("qc-n") { return "qc-n" }
            if lat < 49.0, qcHalves.contains("qc-s") { return "qc-s" }
            return qcHalves[0]
        }
        let caHalves = hits.filter { $0 == "ca-s" || $0 == "ca-n" }
        if !caHalves.isEmpty {
            if lat >= 37.0, caHalves.contains("ca-n") { return "ca-n" }
            if lat < 37.0, caHalves.contains("ca-s") { return "ca-s" }
            return caHalves[0]
        }
        let nlHalves = hits.filter { $0 == "nl-island" || $0 == "nl-lab" }
        if !nlHalves.isEmpty {
            if lon <= -56.8, nlHalves.contains("nl-lab") { return "nl-lab" }
            if lon > -56.8, nlHalves.contains("nl-island") { return "nl-island" }
            return nlHalves[0]
        }
        hits.sort { (bboxes[$0]?.area ?? .infinity) < (bboxes[$1]?.area ?? .infinity) }
        return hits[0]
    }

    /// Alias kept for lockstep with live `maritimesOwner`.
    static func maritimesOwner(longitude lon: Double, latitude lat: Double) -> String? {
        polygonOwner(longitude: lon, latitude: lat)
    }

    private static func geometryBbox(_ geom: [String: Any]) -> BBox? {
        var minLon = Double.infinity
        var minLat = Double.infinity
        var maxLon = -Double.infinity
        var maxLat = -Double.infinity
        func walk(_ raw: Any?) {
            guard let raw else { return }
            if let pair = raw as? [Any], pair.count >= 2,
               let lon = (pair[0] as? NSNumber)?.doubleValue,
               let lat = (pair[1] as? NSNumber)?.doubleValue,
               !(pair[0] is [Any]) {
                if lon < minLon { minLon = lon }
                if lat < minLat { minLat = lat }
                if lon > maxLon { maxLon = lon }
                if lat > maxLat { maxLat = lat }
                return
            }
            if let list = raw as? [Any] {
                for child in list { walk(child) }
            }
        }
        walk(geom["coordinates"])
        guard minLon.isFinite else { return nil }
        return BBox(minLon: minLon, minLat: minLat, maxLon: maxLon, maxLat: maxLat)
    }

    private static func pointInGeometry(lon: Double, lat: Double, type: String, coordinates: Any?) -> Bool {
        if type == "Polygon", let rings = asRings(coordinates) {
            return pointInPolygon(lon: lon, lat: lat, rings: rings)
        }
        if type == "MultiPolygon", let polygons = asMultiRings(coordinates) {
            return polygons.contains { pointInPolygon(lon: lon, lat: lat, rings: $0) }
        }
        return false
    }

    private static func asRings(_ raw: Any?) -> [[[Double]]]? {
        guard let rings = raw as? [Any] else { return nil }
        return rings.compactMap { asRing($0) }
    }

    private static func asMultiRings(_ raw: Any?) -> [[[[Double]]]]? {
        guard let polygons = raw as? [Any] else { return nil }
        return polygons.compactMap { asRings($0) }
    }

    private static func asRing(_ raw: Any) -> [[Double]]? {
        guard let points = raw as? [Any] else { return nil }
        return points.compactMap { point in
            guard let pair = point as? [Any], pair.count >= 2,
                  let lon = (pair[0] as? NSNumber)?.doubleValue,
                  let lat = (pair[1] as? NSNumber)?.doubleValue
            else { return nil }
            return [lon, lat]
        }
    }

    private static func pointInPolygon(lon: Double, lat: Double, rings: [[[Double]]]) -> Bool {
        guard let outer = rings.first, pointInRing(lon: lon, lat: lat, ring: outer) else { return false }
        for hole in rings.dropFirst() where pointInRing(lon: lon, lat: lat, ring: hole) {
            return false
        }
        return true
    }

    private static func pointInRing(lon: Double, lat: Double, ring: [[Double]]) -> Bool {
        var inside = false
        var j = ring.count - 1
        for i in 0..<ring.count {
            let xi = ring[i][0], yi = ring[i][1]
            let xj = ring[j][0], yj = ring[j][1]
            let intersect = (yi > lat) != (yj > lat)
                && lon < (xj - xi) * (lat - yi) / (yj - yi) + xi
            if intersect { inside.toggle() }
            j = i
        }
        return inside
    }
}

private final class BundleToken {}
